### PR TITLE
Refactor of `pairwise_linestring_distance` to use `multilinestring_range`, adds support to multilinestring distance

### LINKED ISSUE #745: [BUG] linestring device code uses std::min when it should use min()
std::min does not map to the GPU hardware min() function. This should be changed.

point_linestring_distance.cuh
linestring_distance.cuh
linestring.cuh

https://github.com/rapidsai/cuspatial/issues/745#tasklist-block-42aba05f-f923-4937-80f3-ae08f46dc762

### PR DESCRIPTION
**Pull request by isVoid**
## Description

Note, this is the first part of `pairwise_linestring_distance` refactoring, part 1 of PR: https://github.com/rapidsai/cuspatial/pull/753

Depends on #752 

Contributes to https://github.com/rapidsai/cuspatial/issues/706, https://github.com/rapidsai/cuspatial/issues/703

Closes #745 

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cuspatial/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/include/cuspatial/experimental/geometry/linestring_ref.cuh [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+#include <cuspatial/cuda_utils.hpp>
+
+namespace cuspatial {
+
+/**
+ * @brief Represent a multipoint stored in structure of array on memory.
```
Comment (thomcom) [inline cpp/include/cuspatial/experimental/geometry/linestring_ref.cuh:22]:
```suggestion
 * @brief Represent a linestring stored in structure of array on memory.
```

File: cpp/include/cuspatial/experimental/detail/ranges/multilinestring_range.cuh [hunk 2]
```diff
@@ -23,15 +23,44 @@
...
+  {
+  }
+
+  CUSPATIAL_HOST_DEVICE auto operator()(difference_type i)
+  {
+    return multilinestring_ref{_part_begin + _geometry_begin[i],
+                               thrust::next(_part_begin + _geometry_begin[i + 1]),
+                               _point_begin,
+                               _point_end};
+  }
```
Comment (thomcom) [inline cpp/include/cuspatial/experimental/detail/ranges/multilinestring_range.cuh:62]: Curious what you're using this for. A constructor that returns a slice of a `multilinestring_range`?

File: cpp/include/cuspatial/experimental/detail/geometry_collection/multilinestring_ref.cuh [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  return _point_begin;
+}
+
+template <typename PartIterator, typename VecIterator>
+CUSPATIAL_HOST_DEVICE auto multilinestring_ref<PartIterator, VecIterator>::point_end() const
+{
+  return _point_end;
+}
+
+}  // namespace cuspatial
```
Comment (thomcom) [inline cpp/include/cuspatial/experimental/detail/geometry_collection/multilinestring_ref.cuh:71]: What is the difference between a `multilinestring_range` and a `multilinestring_ref`?

Comment (isVoid) [reply cpp/include/cuspatial/experimental/detail/geometry_collection/multilinestring_ref.cuh:71]: Conceptually `multilinestring_range` is a non-owning object on an **array** of multilinestrings.

`multilinestring_ref` is just a single multilinestring. Iterating on `multilinestring_range` gives a `multilinestring_ref`.

File: cpp/include/cuspatial/experimental/detail/ranges/multilinestring_range.cuh (hunk 2 above)
Comment (isVoid) [reply cpp/include/cuspatial/experimental/detail/ranges/multilinestring_range.cuh:62]: Which is a multilinestring.

**Review by thomcom — APPROVED** [review]

File: cpp/include/cuspatial/experimental/detail/geometry/linestring_ref.cuh [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  : _point_begin(begin), _point_end(end)
+{
+  using T = iterator_vec_base_type<VecIterator>;
+  static_assert(is_same<vec_2d<T>, iterator_value_type<VecIterator>>(), "must be vec2d type");
+}
+
+template <typename VecIterator>
+CUSPATIAL_HOST_DEVICE auto linestring_ref<VecIterator>::num_segments() const
+{
+  return thrust::distance(_point_begin, _point_end) - 2;
```
Comment (harrism) [inline cpp/include/cuspatial/experimental/detail/geometry/linestring_ref.cuh:54]: Why minus two? Perhaps a one-line comment to answer this.

File: cpp/include/cuspatial/experimental/geometry/linestring_ref.cuh [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+  CUSPATIAL_HOST_DEVICE auto num_segments() const;
+
+  /// Return iterator to the starting point of the multilinestring.
+  CUSPATIAL_HOST_DEVICE auto segment_begin() const;
+  /// Return iterator to one-past the last point of the multilinestring.
+  CUSPATIAL_HOST_DEVICE auto segment_end() const;
+
+  /// Return iterator to the starting point of the multipoint.
+  CUSPATIAL_HOST_DEVICE auto begin() const { return segment_begin(); }
+  /// Return iterator the the one-past the last point of the multipoint.
```
Comment (harrism) [inline cpp/include/cuspatial/experimental/geometry/linestring_ref.cuh:40]:
```suggestion
  /// Return iterator to one past the last point of the multipoint.
```

File: cpp/include/cuspatial/experimental/geometry/linestring_ref.cuh [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+template <typename VecIterator>
+class linestring_ref {
+ public:
+  CUSPATIAL_HOST_DEVICE linestring_ref(VecIterator begin, VecIterator end);
+
+  CUSPATIAL_HOST_DEVICE auto num_segments() const;
+
+  /// Return iterator to the starting point of the multilinestring.
+  CUSPATIAL_HOST_DEVICE auto segment_begin() const;
+  /// Return iterator to one-past the last point of the multilinestring.
```
Comment (harrism) [inline cpp/include/cuspatial/experimental/geometry/linestring_ref.cuh:35]:
```suggestion
  /// Return iterator to one past the last point of the multilinestring.
```

File: cpp/include/cuspatial/experimental/geometry/linestring_ref.cuh [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+#include <cuspatial/cuda_utils.hpp>
+
+namespace cuspatial {
+
+/**
+ * @brief Represent a linestring stored in structure of array on memory.
```
Comment (harrism) [inline cpp/include/cuspatial/experimental/geometry/linestring_ref.cuh:22]:
```suggestion
 * @brief Represent a linestring stored in a structure of arrays.
```

File: cpp/include/cuspatial/experimental/geometry/linestring_ref.cuh [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+  CUSPATIAL_HOST_DEVICE auto segment_begin() const;
+  /// Return iterator to one-past the last point of the multilinestring.
+  CUSPATIAL_HOST_DEVICE auto segment_end() const;
+
+  /// Return iterator to the starting point of the multipoint.
+  CUSPATIAL_HOST_DEVICE auto begin() const { return segment_begin(); }
+  /// Return iterator the the one-past the last point of the multipoint.
+  CUSPATIAL_HOST_DEVICE auto end() const { return segment_end(); }
+
+  VecIterator _point_begin;
```
Comment (harrism) [inline cpp/include/cuspatial/experimental/geometry/linestring_ref.cuh:43]: `protected:`?

File: cpp/include/cuspatial/experimental/geometry_collection/multilinestring_ref.cuh [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+#include <cuspatial/cuda_utils.hpp>
+
+namespace cuspatial {
+
+/**
+ * @brief Represent a multilinestring stored in structure of array on memory.
```
Comment (harrism) [inline cpp/include/cuspatial/experimental/geometry_collection/multilinestring_ref.cuh:22]:
```suggestion
 * @brief Represent a multilinestring stored in a structure of arrays.
```

File: cpp/include/cuspatial/experimental/geometry_collection/multilinestring_ref.cuh [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+ */
+template <typename PartIterator, typename VecIterator>
+class multilinestring_ref {
+ public:
+  CUSPATIAL_HOST_DEVICE multilinestring_ref(PartIterator part_begin,
+                                            PartIterator part_end,
+                                            VecIterator point_begin,
+                                            VecIterator point_end);
+
+  CUSPATIAL_HOST_DEVICE auto num_linestrings() const;
```
Comment (harrism) [inline cpp/include/cuspatial/experimental/geometry_collection/multilinestring_ref.cuh:34]: docs for these two...

File: cpp/include/cuspatial/experimental/geometry_collection/multilinestring_ref.cuh [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+  CUSPATIAL_HOST_DEVICE auto size() const { return num_linestrings(); }
+
+  /// Return iterator to the starting linestring.
+  CUSPATIAL_HOST_DEVICE auto part_begin() const;
+  /// Return iterator to one-past the last linestring.
```
Comment (harrism) [inline cpp/include/cuspatial/experimental/geometry_collection/multilinestring_ref.cuh:39]:
```suggestion
  /// Return iterator to one past the last linestring.
```

File: cpp/include/cuspatial/experimental/geometry_collection/multilinestring_ref.cuh [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+ public:
+  CUSPATIAL_HOST_DEVICE multilinestring_ref(PartIterator part_begin,
+                                            PartIterator part_end,
+                                            VecIterator point_begin,
+                                            VecIterator point_end);
+
+  CUSPATIAL_HOST_DEVICE auto num_linestrings() const;
+  CUSPATIAL_HOST_DEVICE auto size() const { return num_linestrings(); }
+
+  /// Return iterator to the starting linestring.
```
Comment (harrism) [inline cpp/include/cuspatial/experimental/geometry_collection/multilinestring_ref.cuh:37]:
```suggestion
  /// Return iterator to the first linestring.
```

File: cpp/include/cuspatial/experimental/geometry_collection/multilinestring_ref.cuh [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+  CUSPATIAL_HOST_DEVICE auto part_begin() const;
+  /// Return iterator to one-past the last linestring.
+  CUSPATIAL_HOST_DEVICE auto part_end() const;
+
+  /// Return iterator to the starting point of the multipoint.
```
Comment (harrism) [inline cpp/include/cuspatial/experimental/geometry_collection/multilinestring_ref.cuh:42]:
```suggestion
  /// Return iterator to the first point of the multilinestring.
```

File: cpp/include/cuspatial/experimental/geometry_collection/multilinestring_ref.cuh [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+  CUSPATIAL_HOST_DEVICE auto point_begin() const;
+  /// Return iterator to one-past the last point of the multipoint.
```
Comment (harrism) [inline cpp/include/cuspatial/experimental/geometry_collection/multilinestring_ref.cuh:44]:
```suggestion
  /// Return iterator to one past the last point of the multilinestring.
```

File: cpp/include/cuspatial/experimental/geometry_collection/multilinestring_ref.cuh [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+  CUSPATIAL_HOST_DEVICE auto point_end() const;
+
+  /// Return iterator to the starting point of the multipoint.
```
Comment (harrism) [inline cpp/include/cuspatial/experimental/geometry_collection/multilinestring_ref.cuh:47]:
```suggestion
  /// Return iterator to the first linestring of the multilinestring.
```

File: cpp/include/cuspatial/experimental/geometry_collection/multilinestring_ref.cuh [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+  CUSPATIAL_HOST_DEVICE auto begin() const { return part_begin(); }
+  /// Return iterator the the one-past the last point of the multipoint.
```
Comment (harrism) [inline cpp/include/cuspatial/experimental/geometry_collection/multilinestring_ref.cuh:49]:
```suggestion
  /// Return iterator the the one past the last linestring of the multilinestring.
```

File: cpp/include/cuspatial/experimental/ranges/multilinestring_range.cuh [hunk 17]
```diff
@@ -83,6 +83,14 @@ class multilinestring_range {
    */
   CUSPATIAL_HOST_DEVICE auto num_points();
 
+  CUSPATIAL_HOST_DEVICE auto multilinestring_begin();
```
Comment (harrism) [inline cpp/include/cuspatial/experimental/ranges/multilinestring_range.cuh:86]: docs for these four.

File: cpp/include/cuspatial/experimental/ranges/multilinestring_range.cuh [hunk 18]
```diff
@@ -118,13 +126,16 @@ class multilinestring_range {
   CUSPATIAL_HOST_DEVICE thrust::pair<vec_2d<element_t>, vec_2d<element_t>> segment(
     IndexType segment_idx);
 
+  template <typename IndexType>
```
Comment (harrism) [inline cpp/include/cuspatial/experimental/ranges/multilinestring_range.cuh:129]: docs

File: cpp/include/cuspatial/experimental/geometry_collection/multipoint_ref.cuh [hunk 19]
```diff
@@ -19,7 +19,7 @@
 namespace cuspatial {
 
 /**
- * @brief Represent a multipoint stored in structure of array on memory.
+ * @brief Represent a reference to multipoint stored in structure of array on memory.
```
Comment (harrism) [inline cpp/include/cuspatial/experimental/geometry_collection/multipoint_ref.cuh:22]: Also add this "a reference to" to all the other `*_ref` docstrings.

```suggestion
 * @brief Represent a reference to a multipoint stored in a structure of arrays.
```

File: cpp/include/cuspatial/experimental/linestring_distance.cuh [hunk 20]
```diff
@@ -28,48 +28,22 @@ namespace cuspatial {
...
- * @tparam Cart2dItA iterator type for point array of the first linestring of each pair. Must meet
- * the requirements of [LegacyRandomAccessIterator][LinkLRAI] and be device-accessible.
- * @tparam Cart2dItB iterator type for point array of the second linestring of each pair. Must meet
- * the requirements of [LegacyRandomAccessIterator][LinkLRAI] and be device-accessible.
- * @tparam OffsetIterator iterator type for offset array. Must meet the requirements of
- * [LegacyRandomAccessIterator][LinkLRAI] and be device-accessible.
- * @tparam OutputIt iterator type for output array. Must meet the requirements of
- * [LegacyRandomAccessIterator][LinkLRAI] and be device-accessible.
+ * @tparam MultiLinestringRange an instance of template type `multilinestring_range`
+ * @tparam OutputIt iterator type for output array. Must meet the requirements of [LRAI](LinkLRAI).
```
Comment (harrism) [inline cpp/include/cuspatial/experimental/linestring_distance.cuh:32]: and be device writeable?

File: cpp/src/spatial/linestring_distance.cu [hunk 21]
```diff
@@ -14,27 +14,19 @@
  * limitations under the License.
  */
 
+#include "thrust/iterator/counting_iterator.h"
```
Comment (harrism) [inline cpp/src/spatial/linestring_distance.cu:17]: auto-added include. angle brackets and move to thrust section.

File: cpp/tests/experimental/spatial/linestring_distance_test.cu [hunk 22]
```diff
@@ -19,23 +19,32 @@
...
 #include <thrust/iterator/counting_iterator.h>
 
+#include <initializer_list>
+
 namespace cuspatial {
 namespace test {
 
-using namespace cudf;
+template <typename T>
+auto make_device_vector(std::initializer_list<T> inl)
```
Comment (harrism) [inline cpp/tests/experimental/spatial/linestring_distance_test.cu:36]: Ooh, this looks handy. Perhaps it could be moved to a test utility header...

**Review by harrism — COMMENTED** [review]
Looks very nice. Various small doc suggestions only.

File: cpp/include/cuspatial/experimental/detail/geometry/linestring_ref.cuh (hunk 4 above)
Comment (isVoid) [reply cpp/include/cuspatial/experimental/detail/geometry/linestring_ref.cuh:54]: 🤔

Comment (isVoid) [reply cpp/include/cuspatial/experimental/detail/geometry/linestring_ref.cuh:54]: It's incorrect. Should be -1. `to_linestring_functor` incorrectly computed the start and end iterator for points. Leading to an extra count of point. The number of segments is the number of points -1. The number of points is exactly `thrust::distance(_point_begin, _point_end)`.

File: cpp/tests/experimental/spatial/linestring_distance_test.cu (hunk 22 above)
Comment (isVoid) [reply cpp/tests/experimental/spatial/linestring_distance_test.cu]: Done

File: cpp/include/cuspatial/experimental/detail/geometry/linestring_ref.cuh (hunk 4 above)
Comment (thomcom) [reply cpp/include/cuspatial/experimental/detail/geometry/linestring_ref.cuh]: Haha good find. :D

**Comment by isVoid** [comment]
rerun tests

**Comment by isVoid** [comment]
rerun tests

**Review by harrism — APPROVED** [review]
Looks good!

**Comment by harrism** [comment]
@gpucibot merge

**Pull request merged**